### DISCUSSION TITLE
make install 似乎执行了cp bin/binary bin

### DISCUSSION THREAD

**Discussion by DanielKui** [xmake-io/xmake#4013] [Q&A] [answered]
includes ("../projCompileOption.lua")
includes ("../Tools/LogManager/LogManager.lua")
includes ("HXQuoteDemoFiles.lua")

add_rules("mode.debug", "mode.release")

target ("HXQuoteApi")
	add_packages("fmt")
	set_optimize("faster")
	set_languages(cppVersion)
	add_ldflags(linkOptions)
    set_kind ("shared")
	set_targetdir(tarDir)
	set_installdir(".")
	add_ldflags("-Wl,-rpath="..tarDir)
	add_deps(logdeps)
    add_defines (ApiDefine)
	add_includedirs(ApiSrcDir)
	add_files(ApiCC)
    add_linkdirs(ApiLibDir)
    add_links(ApiLib)
    
	after_install(function (target)
		print ("target:targetdir():%s",target:targetdir())
		print ("installdir():%s",target:installdir())
		print ("installdir():%s",target:installdir().."/lib/*")
		print ("installdir():%s",target:installdir().."/bin/*")
		
		print ("os.projectdir():%s",os.projectdir())
		
		if is_mode("release") then
			os.trycp(path.join(os.projectdir(),"/Tools/LogManager/bin/libLogManager_x64.so"), target:targetdir(),{symlink = true})
		end
		if is_mode("debug") then
			os.trycp(path.join(os.projectdir(),"/Tools/LogManager/bin/libLogManagerd_x64.so"), target:targetdir(),{symlink = true})
		end
		os.trycp(path.join(os.projectdir(),"/ThirdParty/lib64_2022/libquill.so*"), target:targetdir(),{symlink = true})
		os.trycp(path.join(os.projectdir(),"/ThirdParty/lib64_2022/libprotobuf-lite.so.23"), target:targetdir(),{symlink = true})
		os.trycp(path.join(os.projectdir(),"/ThirdParty/lib64_2022/libACE.so.6.5.9"), target:targetdir(),{symlink = true})

		os.trycp(target:installdir().."/lib/*", target:targetdir(),{symlink = true})
    end)
	
	if is_mode("debug") then
		add_defines("DEBUG")
		set_basename("HXQuoteApid_x64")
		set_symbols("debug")
		set_optimize("none")
		add_cxxflags(debugOptions)
		add_cflags(pubCRule)
	end
	if is_mode("release") then
		add_defines("NDEBUG")
		set_basename("HXQuoteApi_x64")
		set_optimize("faster")
		add_cxxflags(releaseOptions,{force=true})
		add_cflags(pubCRule)
	end
	
	on_clean(function (target)
        -- Delete only target files
        os.rm(target:targetfile())
    end)

target ("HXQuoteDemo")
	add_packages("fmt")
	set_kind ("binary")
	set_languages(cppVersion)
	add_ldflags(linkOptions)    
	set_targetdir(tarDir)
	set_installdir(".")
    add_defines (DemoDefine)
	
	add_includedirs(DemoSrcDir)
	add_files(DemoCC)
    add_linkdirs(DemoLibDir)
    add_links(DemoLib)
	add_deps(hxquotedeps)
	on_clean(function (target)
        -- Delete only target files
        os.rm(target:targetfile())
    end)

	if is_mode("debug") then
		add_defines("DEBUG")
		set_basename("HXQuoteDemod_x64")
		set_symbols("debug")
		set_optimize("none")
		add_cxxflags(debugOptions)
		add_cflags(pubCRule)
		add_links("HXQuoteApid_x64")
	end
	if is_mode("release") then
		add_defines("NDEBUG")
		set_basename("HXQuoteDemo_x64")
		set_optimize("faster")
		add_links("HXQuoteApi_x64")
		add_cxxflags(releaseOptions,{force=true})
		add_cflags(pubCRule)
	end

![image](https://github.com/xmake-io/xmake/assets/578118/37f8059a-47b9-4aed-b5d4-06e8236d37bc)

我执行了xmake后，进行xmake install， 结果HXApi/bin/HXQuoteDemo_x64这个binary的size变成0了，xmake后我看到是正常的，但xmake install后就变成0了

**Comment by waruqi** [comment]
> os.trycp(target:installdir().."/lib/*", target:targetdir(),{symlink = true})

target:installdir() 是 target 的安装目录，安装时候，拿安装目录作为源路径，空目录 copy 么？

**Comment by DanielKui** [comment]
![image](https://github.com/xmake-io/xmake/assets/578118/7a3d7403-3a57-43fd-816c-ca0ef48336a0)
这是我的installdir的目录内容，应该是的有文件的

**Comment by DanielKui** [comment]
![image](https://github.com/xmake-io/xmake/assets/578118/8aba46bf-f4e3-4a5b-87b3-2d7f692cbdbc)


我是用了set_installdir(".") 

HXApi/lib是存放 fmt.so的位置（add_packages拷贝过去的），我想把他们拷贝到HXApi/bin下

**Reply by waruqi** [reply]
看不懂你的搞法，反正你自己 print 调调么，这种又不是很难调。。自己试个几次就出来了。。

**Comment by DanielKui** [comment]
谢谢答复，请教下，  add_packages("fmt")下载编译后后，放到~/.xmake/packagtes/xxxxx/libfmt.so目录下了，我如何在on_install里获取到这个libfmt.so的所在目录呢？因为目录结构里有一串随机变量

**Reply by waruqi** [reply, marked as answer]
多翻翻源码 https://github.com/xmake-io/xmake/blob/master/xmake/modules/target/action/install/unix.lua
